feat(create): add 'copy from existing' (clone) option to CreateComponent

sogladev opened the pull request:
### Changed proposed:

Add support for cloning an existing entity from the 'Create' card for multiple types.

I modified the tests for `create.component.spec.ts`. I'm not sure what other tests are needed for this.

FYI not an NX/Angular expert and I used AI tools to help draft this. Hopefully, nothing too bad in here 

### Details:

Adds `allowCopy` field to the CreateComponent `<keira-create>`, which is opt-in.

Handling is done by routing to a `/clone` route with a provided `ID` to copy from. This is similar to `/select` and `/create` that already exist. The `/copy` route displays a page with `copy` queries for the main table and related tables. After executing, the user is routed to the edit page.

After doing this for creatures, it was easy enough to copy and paste the handling for other types. 

For gobjects and creatures, only the template-like tables are copied. SmartAI (TODO?), spawns or guid-specific stuff is not copied.

### Issues

- Closes https://github.com/azerothcore/Keira3/issues/25
- Closes https://github.com/azerothcore/Keira3/issues/700
- Closes https://github.com/azerothcore/Keira3/issues/964


### UI 


<details> <summary> UI Screenshots </summary>



![keira_38_1](https://github.com/user-attachments/assets/a60fce27-3e1e-4bed-aba8-490638cc9302)

![keira_3](https://github.com/user-attachments/assets/f630e228-f3c2-4be2-9e31-b57993af6a0e)


![keira_4](https://github.com/user-attachments/assets/f2b47555-3abd-4fbb-9715-2dfe57874022)




</details>


### [[copilot summary below]]

This pull request adds support for "copy" routes and features across multiple entity types in the Keira app, allowing users to duplicate existing entries. It also introduces new internationalized strings related to the copy functionality in all supported languages.

Key changes include:

**Routing and Feature Enhancements:**

* Added new "copy" routes for various entities (e.g., Creature, Quest, Gameobject, Item, Spell, Gossip, GameTele, FishingLoot, MailLoot, ReferenceLoot, PageText, NpcText, AcoreString) in `routes.ts`, each guarded by the appropriate handler service and linked to the corresponding `CopyComponent`. This enables users to copy existing records for these entities. [[1]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR131-R135) [[2]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR245-R249) [[3]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR304-R308) [[4]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR353-R357) [[5]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR397-R401) [[6]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR420-R424) [[7]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR434-R438) [[8]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR448-R452) [[9]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR471-R475) [[10]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR486-R490) [[11]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR504-R508) [[12]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR564-R568) [[13]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR578-R582)
* Updated imports in `routes.ts` to include all new `CopyComponent` classes for the affected features. [[1]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR7) [[2]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR32) [[3]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fL40-R53) [[4]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR65-R67) [[5]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR89-R92) [[6]](diffhunk://#diff-07dbda67aad6f84c54a7888b8fb62fea11e442aa1eb8c5b1c9ecb1010414071fR102-R112)

**Internationalization (i18n):**

* Added new strings for copy-related UI elements (such as "Copy from existing", "Copy Entry SQL", "Execute", "Continue to Editor", etc.) to all supported language files (`en.json`, `de.json`, `es.json`, `fr.json`, `el.json`, `it.json`). This ensures the new copy functionality is fully localized. [[1]](diffhunk://#diff-03790dfa92a094be04604084f36040ca87d0e4c89903ebb0d7106f8ea1102e06L28-R33) [[2]](diffhunk://#diff-03790dfa92a094be04604084f36040ca87d0e4c89903ebb0d7106f8ea1102e06R218-R230) [[3]](diffhunk://#diff-9d271ee9599b6b8312b8f5db49bab343d8e9ac8d9e07a674423cb2d0102fe2efL28-R33) [[4]](diffhunk://#diff-9d271ee9599b6b8312b8f5db49bab343d8e9ac8d9e07a674423cb2d0102fe2efR218-R230) [[5]](diffhunk://#diff-e26b26b546bc56fc6d7c0d75190bf0847276b837abf9e369bfc960ca9a1aa875L28-R33) [[6]](diffhunk://#diff-e26b26b546bc56fc6d7c0d75190bf0847276b837abf9e369bfc960ca9a1aa875R218-R230) [[7]](diffhunk://#diff-319fddd3dbf3e557304fab9ea9e0cb2c31011fa957613d15b21653646ab5dd9eL28-R33) [[8]](diffhunk://#diff-319fddd3dbf3e557304fab9ea9e0cb2c31011fa957613d15b21653646ab5dd9eR218-R230) [[9]](diffhunk://#diff-fb91f9727bacb5a56884440682a97a99afc59f8a3baf22b01b85b47916502666L28-R33) [[10]](diffhunk://#diff-fb91f9727bacb5a56884440682a97a99afc59f8a3baf22b01b85b47916502666R218-R230) [[11]](diffhunk://#diff-5cad24643c18c777662b78ac04a6cc95964ce0df53bfa9d8148303e33d910276L28-R33)

These changes collectively introduce and localize the experimental ability to copy existing entries for a wide range of entities within the Keira application, improving user productivity and data management.

sogladev committed:
feat(create): copy from existing entry

Helias commented on libs/features/creature/src/creature-copy/creature-copy.component.ts:
```diff
@@ -0,0 +1,77 @@
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  selector: 'keira-creature-copy',
+  templateUrl: './creature-copy.component.html',
+  standalone: true,
```
this line can be removed, by default standalone is true

Helias commented on libs/features/game-tele/src/index.ts:
```diff
@@ -1,3 +1,4 @@
 export { GameTeleComponent } from './edit-game-tele/game-tele.component';
 export { SelectGameTeleComponent } from './select-game-tele/select-game-tele.component';
 export { GameTeleHandlerService } from './game-tele-handler.service';
+export * from './game-tele-copy/game-tele-copy.component';
```
I would export the single entities instead of `*`

Helias commented on libs/features/other-loots/src/spell-loot/spell-loot-handler.service.ts:
```diff
@@ -7,6 +7,16 @@ import { SPELL_LOOT_TEMPLATE_TABLE, SpellLootTemplate } from '@keira/shared/acor
+      this.router.navigate([this.copyRoutePath]);
+    } else {
+      super.select(isNew, id, name, navigate);
+    }
```
this implementation is always the same in all files? it seems so, we could move this implementation in the upper class `HandlerService` defining only `copyRoutePath` here or moving it into the class `SpellLootTemplate` (?) 

What do you think?

Helias commented on libs/shared/base-editor-components/src/create/create.component.ts:
```diff
@@ -21,13 +21,32 @@ export class CreateComponent<T extends TableRow> extends SubscriptionHandler imp
+  }
+  get allowCopy(): boolean {
+    return this._allowCopy;
+  }
```
Not strictly necessary, I could refactor it a bit later, but I would use signals for this if possible

Helias commented on libs/shared/base-editor-components/src/create/create.component.html:
```diff
@@ -1,37 +1,103 @@
+          [(ngModel)]="creationMethod"
+          (change)="onCreationMethodChange()"
+          [disabled]="!allowCopy"
+          [title]="allowCopy ? '' : 'Copy from existing is not available for this entity'"
```
you could set this text in the translation file

Helias commented on libs/shared/base-editor-components/src/copy-output/copy-output.component.ts:
```diff
@@ -0,0 +1,312 @@
+    this.populateRelatedTables();
+  }
+
+  protected setCopyModeForTable(tableName: string, mode: string): void {
```
mode could have the type raw/all

Helias commented on libs/features/texts/src/index.ts:
```diff
@@ -9,7 +10,9 @@ export { BroadcastTextComponent } from './broadcast-text/broadcast-text.componen
 export { NpcTextHandlerService } from './npc-text/npc-text-handler.service';
 export { SelectNpcTextComponent } from './npc-text/select-npc-text.component';
 export { NpcTextComponent } from './npc-text/npc-text.component';
+export * from './npc-text/npc-text-copy.component';
```
export the single component not all

Helias commented:
It's a big job, it's fantastic :100: , I left some comments

sogladev committed:
update with suggestions

remove default

specify imports

remove some imports

move select handler logic to shared handler

move text to translation file

remove 'public'

add 'protected'

refactor @Input to signal, define interface

remove 'standalone'

actions-user committed:
fix: missing translation keys

Helias committed:
Merge branch 'master' into dev-copy-from

sogladev committed:
delegate copy route select override to super HandlerService

sogladev committed:
test handler route

mock copyRoutePath to allow test to override it

sogladev committed:
add coverage for mysql copy

sogladev committed:
fix 'Property 'checkId' is protected' in test

sogladev committed:
add coverage for create.component

FrancescoBorzi commented on libs/shared/base-editor-components/src/copy-output/copy-output.component.ts:
```diff
@@ -0,0 +1,319 @@
+    // Navigate to the editor for the newly created entry
+    this.handlerService()!.select(false, this.newId());
+  }
+}
```
bonus: maybe some of the logic of these methods can be moved to some pure functions or to a service? but not mandatory given that it's not always really doable

sogladev committed:
refactor: use signals for create component

sogladev committed:
refactor: apply suggestion to remove super

sogladev committed:
refactor: extract type and interface to model file

sogladev committed:
refactor: avoid using shared mutable state between tests

sogladev committed:
fix: coverage back to 100%

sogladev committed:
refactor: extract logic to service

sogladev committed:
fix: lint

sogladev committed:
fixup! refactor: use signals for create component

sogladev committed:
Merge branch 'master' into dev-copy-from

sogladev committed:
fix lint
